src/dataflow/source/kinesis.rs

```diff
@@ -82,78 +83,96 @@ where
             // pause execution and reactivate later.
             let timer = std::time::Instant::now();
 
-            // When the next_shard_iterator is null, the shard has been closed and the
-            // requested iterator does not return any more data.
-            while let Some(iterator) = &shard_iterator {
-                let get_records_output = match block_on(get_records(&client, &iterator)) {
-                    Ok(output) => {
-                        *shard_iterator = output.next_shard_iterator.clone();
-                        output
-                    }
-                    Err(RusotoError::HttpDispatch(e)) => {
-                        // todo@jldlaughlin: Parse this to determine fatal/retriable?
-                        error!("{}", e);
-                        activator.activate_after(get_reactivation_duration(timer));
-                        return SourceStatus::Alive;
-                    }
-                    Err(RusotoError::Service(GetRecordsError::ExpiredIterator(e))) => {
-                        // todo@jldlaughlin: Will need track source offsets to grab a new iterator.
-                        error!("{}", e);
-                        return SourceStatus::Done;
-                    }
-                    Err(RusotoError::Service(GetRecordsError::ProvisionedThroughputExceeded(
-                        _,
-                    ))) => {
-                        activator.activate_after(get_reactivation_duration(timer));
-                        return SourceStatus::Alive;
-                    }
-                    Err(e) => {
-                        // Fatal service errors:
-                        //  - InvalidArgument
-                        //  - KMSAccessDenied, KMSDisabled, KMSInvalidState, KMSNotFound,
-                        //    KMSOptInRequired, KMSThrottling
-                        //  - ResourceNotFound
-                        //
-                        // Other fatal Rusoto errors:
-                        // - Credentials
-                        // - Validation
-                        // - ParseError
-                        // - Unknown (raw HTTP provided)
-                        // - Blocking
-                        error!("{}", e);
-                        return SourceStatus::Done;
-                    }
-                };
+            // Rotate through the queue of shards. Start with a new
+            // shard on each activation.
+            while let Some(shard_id) = shard_queue.pop_front() {
+                // While the next_shard_iterator is Some(iterator), the shard is open
+                // and could return more data.
+                while let Some(iterator) = shard_to_iterator.get(&shard_id).unwrap_or_else(|| {
```

Comment:
I might be misreading here but won't the second while let cause you to always read from the same shard as long as it's open (preventing you from  reading from another shard) until you reach line 144 (the > 10 ms  if statement) at which point you'll get booted out of the function, and on the next activation only will you start reading from the second shard? Do you not want an if let Some(iterator) there?